mlflow/store/tracking/sqlalchemy_store.py

```diff
@@ -457,30 +457,81 @@ def _get_deleted_runs(self):
+
+    def log_metrics(self, run_id, metrics, auto_new=True):
+        if metrics is None:
+            return
+
         with self.ManagedSessionMaker() as session:
-            run = self._get_run(run_uuid=run_id, session=session)
-            self._check_run_is_active(run)
-            # ToDo: Consider prior checks for null, type, metric name validations, ... etc.
-            logged_metric, just_created = self._get_or_create(
-                model=SqlMetric, run_uuid=run_id, key=metric.key, value=value,
-                timestamp=metric.timestamp, step=metric.step, session=session, is_nan=is_nan)
-            # Conditionally update the ``latest_metrics`` table if the logged metric  was not
-            # already present in the ``metrics`` table. If the logged metric was already present,
-            # we assume that the ``latest_metrics`` table already accounts for its presence
-            if just_created:
-                self._update_latest_metric_if_necessary(logged_metric, session)
+            try:
+                run = self._get_run(run_uuid=run_id, session=session)
+                self._check_run_is_active(run)
+
+                metrics_per_key = {}
+
+                for metric in metrics:
+
+                    _validate_metric(metric.key, metric.value, metric.timestamp, metric.step)
+                    is_nan = math.isnan(metric.value)
+                    if is_nan:
+                        value = 0
+                    elif math.isinf(metric.value):
+                        #  NB: Sql can not represent Infs = > We replace +/- Inf
+                        #  with max/min 64b float value
+                        value = 1.7976931348623157e308 if metric.value > 0 \
+                            else -1.7976931348623157e308
+                    else:
+                        value = metric.value
+
+                    # ToDo: Consider prior checks for null, type, metric name validations, ... etc.
+
+                    # Entries will be added to session. If conflicts are detection during saving,
+                    # metrics are inserted one after another using _get_or_create
+                    if auto_new:
+                        logged_metric = SqlMetric(run_uuid=run_id, key=metric.key, value=value,
+                                                  timestamp=metric.timestamp, step=metric.step,
+                                                  is_nan=is_nan)
+                        self._save_to_db(session, logged_metric)
+                        just_created = True
+
+                    else:
+                        logged_metric, just_created = self._get_or_create(
+                            model=SqlMetric, session=session, run_uuid=run_id, key=metric.key,
+                            value=value, timestamp=metric.timestamp, step=metric.step,
+                            is_nan=is_nan)
+
+                    if just_created:
+                        if metric.key not in metrics_per_key:
```

Comment:
these conditions sound contradictory. They are prolly required due to shared session. Please add a comment to clarify that

Reply:
Also please add a comment on why _all_ metrics are being grouped by the metric name -- a bit hard for reader to follow until they see call to `_update_latest_metric_if_necessary`

Reply:
I added a comment in the code for that. Regarding your first comment, I not sure I understand it correctly. In line 504, it will be checked if the metric is new (this might be not the case, if batch mode is disabled. In this case the metric with same step, value, ts, and name is already stored). Line 505: the metrics are store in a dictionary with the name as the key. If the key does not exist, an empty list will be added to the dictionary to collect all metrics.